docs/madr/decisions/046-meshservice-policy-matching.md

```diff
@@ -0,0 +1,313 @@
+```
+
+The `name` of a `MeshService` ref always refers to the name of one, real Kubernetes
+object in the zone of the policy.
+
+Exactly one of either `targetRef.name` or `targetRef.labels` is **required**.
+
+#### `sectionName`
+
+If `sectionName` is set in `targetRef.kind` with `kind: MeshService`,
+it refers to an entry in `ports` by name and only traffic to that port is affected.
+
+In `backendRefs`, `port` is required. There is no `sectionName`, in accordance
+with Gateway API HTTPRoute.
+
+#### `name`
+
+If the policy ref sets `name`, it refers to one `MeshService` in the *local
+zone*.
+
+On a k8s zone:
+* without `namespace`, the policy refers to `MeshServices` in the same
+  namespace as the policy.
+* with `namespace`, only that namespace is searched for matching `MeshServices`
+
+On a universal zone, `namespace` is not valid.
+
+Note that this means it's possible to refer to synced `MeshServices` via their
+transformed `<service>-<hash-suffix>` name.
+
+#### `labels`
+
+If `name` is not set, `labels` must be set and `namespace` cannot be set!
+
+Via `labels`, groups of `MeshServices` can be matched, for example by:
+
+* `kuma.io/display-name`
+* `kuma.io/zone`
+
+With `labels`, we don't assume anything about the namespace of the object. The
+namespace _must_ be selected using the `k8s.kuma.io/namespace` label.
+
+This also means more verbosity for the headless service usecase.
+It means we don't infer any similarity between the local namespace
+and the other zone's namespace.
+
+#### Examples
+
+These examples assume the `frontend` namespace and the zone named `local-zone`.
+
+##### `MeshService`
+
+Users can target `MeshServices` in the namespace of the policy
+directly with `name`:
+
+```yaml
+spec:
+ to:
+ - targetRef:
+     kind: MeshService
+     name: backend
+```
+
+or by using `labels` without `name`. The below examples search _any namespace_.
+
+```yaml
+spec:
+ to:
+ - targetRef:
+     kind: MeshService
+     labels:
+       k8s.kuma.io/service-name: zk
+       k8s.kuma.io/namespace: zk-namespace # this must be set to target a namespace, even the local one
+       kuma.io/zone: local-zone # must also be set to target a local zone, otherwise all zones are targeted
+```
+
+```yaml
+spec:
+ to:
+ - targetRef:
+     kind: MeshService
+     labels:
+        kuma.io/display-name: backend
+        kuma.io/zone: other-zone
+```
+
+
+##### `backendRef`
+
+The structure of `backendRef` for `MeshService` is largely similar except that
+we *require* a specific port.
+
+```yaml
+kind: MeshHTTPRoute
+spec:
+  targetRef:
+    kind: MeshGateway
+    name: edge-gateway
+  to:
+   - targetRef:
+       kind: Mesh
+     rules:
+       - matches:
+           - path:
+               type: Prefix
+               value: /v1
+         default:
+           backendRefs:
+             - kind: MeshService
+               port: 8080
+               labels:
+                 kuma.io/display-name: frontend
+                 kuma.io/zone: east
```

Comment:
And in particular, what happens to `weight`? Not sure...

Reply:
This really should only be able to match one specific service/port combo. I think we need to validate this to be the case.

Reply:
but we can validate this only when the policy is applied (not when it's created) right?

Reply:
First of all, we can require `display-name` and `zone` labels.

The only problem is if we have a target Kubernetes zone and don't include `namespace`. In that case, we can treat the ref as invalid no matter how many matching services really exist. Ideally this would be reported as an error condition

Reply:
I've updated MADR with: `backendRef cannot match more than one MeshService. When backendRef matches multiple MeshServices this should result in error, and route should be applied as if no services were matched.`
this should cover the issue